### PR TITLE
[opencl] fixed linkage when libOpenCL is static

### PR DESCRIPTION
**Pull request by ilya-lavrenov**
**Problem:** when OpenCL is built as static library, but used with `BUILD_SHARED_LIBS=ON`, then `${CMAKE_DL_LIBS}` are not added to linkage and we have undefined symbols.

**Solution:** use `VCPKG_LIBRARY_LINKAGE` of OpenCL port to "save" library type and not rely on user's `BUILD_SHARED_LIBS`

### CODE REVIEW

**Comment by Cheney-W** [comment]
Waiting https://github.com/microsoft/vcpkg/pull/31627 is merged.

File: ports/opencl/vcpkg-cmake-wrapper.cmake [hunk 1]
```diff
@@ -1,6 +1,20 @@
 _find_package(${ARGS})
-if(NOT BUILD_SHARED_LIBS AND (NOT CMAKE_SYSTEM_NAME MATCHES "Darwin"))
-  find_package(Threads REQUIRED)
+
+include(CMakeFindDependencyMacro)
```
Comment (dg0yt) [inline ports/opencl/vcpkg-cmake-wrapper.cmake:3]: This - and its provided `find_dependency` - cannot be used in wrappers or find modules, only in config files. They rely on not being in the scope of the caller, but the wrapper is called in that scope.

File: ports/opencl/vcpkg-cmake-wrapper.cmake [hunk 2]
```diff
@@ -1,6 +1,20 @@
 _find_package(${ARGS})
-if(NOT BUILD_SHARED_LIBS AND (NOT CMAKE_SYSTEM_NAME MATCHES "Darwin"))
-  find_package(Threads REQUIRED)
+
+include(CMakeFindDependencyMacro)
+
+if(TARGET OpenCL::OpenCL)
+  get_target_property(OpenCL_Shared_Library OpenCL::OpenCL TYPE)
+  if(OpenCL_Shared_Library STREQUAL "SHARED_LIBRARY")
```
Comment (dg0yt) [inline ports/opencl/vcpkg-cmake-wrapper.cmake:7]: This doesn't need to be inspected when the wrapper is used.
It should be burned into the wrapper when the port is built, i.e.
~~~cmake
if("@VCPKG_LIBRARY_LINKAGE@" STREQUAL "dynamic")
~~~
here, plus `configure_file(... wrapper ...)` in the portfile.

**Review by dg0yt — CHANGES_REQUESTED** [review]

File: ports/opencl/vcpkg-cmake-wrapper.cmake (hunk 2 above)
Comment (dg0yt) [reply ports/opencl/vcpkg-cmake-wrapper.cmake:7]: PS: Basically this, applied as replacement for `${BUILD_SHARED_LIBS}` in the original file.

**Review by Cheney-W — APPROVED** [review]

**Comment by ilya-lavrenov** [comment]
Can we merge the PR?

**Pull request merged**